### ISSUE TITLE
ParamScheduler doesn't support multiple optimizers/multiple parameter groups

### ISSUE THREAD

**Issue by amitibo** [pytorch/ignite#317]
Currently `ParamScheduler` (in the `contrib` module) does not support per-parameter scheduling (https://pytorch.org/docs/stable/optim.html), i.e. the scheduled parameter will be updated uniformly in all the parameter groups of the optimizer. It will be beneficial to be able to update only a specific parameter group.

It also doesn't fully support the use of multiple optimizers. For example if two param_schedulers will be set to update the same parameter of two different optimizers. Both will write to the same `param_history` field stored in the `engine.state`.

Will appreciate any ideas on how to extend the `ParamScheduler` API to support both these cases.

**Comment by amitibo** [comment]
A possible solution is to make the parameters more general, i.e. leave it to the user to decide what to do with their value. For example the user can specify a callback that sets a specific parameter group or a parameter of a specific optimizer.
Parameters should also be given a name (similar to `Metric`) and this name should be used for keeping track of their history. This way, different `ParamScheduler`s that update the same parameter of multiple optimizers will not overwrite the same field in the `engine.state.param_history`.

**Comment by amitibo** [comment]
Another "less radical" solution is instead to allow the option of passing a specific `optimizer.param_groups` item (as the `optimizer` argument).

**Comment by vfdev-5** [comment]
@amitibo thanks for the issue, seems interesting idea is raised. I would like to understand better your proposition, if I well understand.

Let's say my model is composed of `stem`, `blocks` and `classifier`.

> Currently ParamScheduler (in the contrib module) does not support per-parameter scheduling 
(https://pytorch.org/docs/stable/optim.html), i.e. the scheduled parameter will be updated uniformly in all the parameter groups of the optimizer. It will be beneficial to be able to update only a specific parameter group.

So you would like to be able to apply different param schedules for each group, something like this ?
```
optim1 = SGD(model.stem.parameters(), 'lr': 0.0001)
optim2 = SGD(model.blocks.parameters(), 'lr': 0.001)
optim3 = SGD(model.classifier.parameters(), 'lr': 0.01)

scheduler1 = LinearCyclicalScheduler(optim1, 'lr', 1e-7, 1e-5, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler1)

scheduler2 = LinearCyclicalScheduler(optim2, 'lr', 1e-5, 1e-3, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler2)

scheduler3 = LinearCyclicalScheduler(optim3, 'lr', 1e-3, 1e-1, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler3)
```

**Comment by amitibo** [comment]
@vfdev-5 my proposition handles two cases:

### multiple optimizers
```
optim1 = SGD(model.stem.parameters(), 'lr': 0.0001)
optim2 = SGD(model.blocks.parameters(), 'lr': 0.001)
optim3 = SGD(model.classifier.parameters(), 'lr': 0.01)

scheduler1 = LinearCyclicalScheduler(optim1, 'lr', 1e-7, 1e-5, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler1, "stem lr")

scheduler2 = LinearCyclicalScheduler(optim2, 'lr', 1e-5, 1e-3, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler2, "blocks lr")

scheduler3 = LinearCyclicalScheduler(optim3, 'lr', 1e-3, 1e-1, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler3, "classifier lr")
```

### parameter groups
```
optim = SGD(
    [
        {"params": model.stem.parameters(), 'lr': 0.001),
        {"params": model.blocks.parameters(), 'lr': 0.01),
        {"params": model.classifier.parameters(), 'lr': 0.0001)
    ]
)

scheduler1 = LinearCyclicalScheduler(optim.param_groups[0], 'lr', 1e-7, 1e-5, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler1, "stem lr")

scheduler2 = LinearCyclicalScheduler(optim.param_groups[1], 'lr', 1e-5, 1e-3, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler2, "blocks lr")

scheduler3 = LinearCyclicalScheduler(optim.param_groups[2], 'lr', 1e-3, 1e-1, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler3, "classifier lr")
```

**Comment by vfdev-5** [comment]
@amitibo the first case I think is already covered just by defining multiple optimizers, no need of "stem lr", "blocks lr" etc.

The second case is very similar to the first case and just avoids defining multiple optimizers and we can reuse simple engines with a single optimizer.

**Comment by amitibo** [comment]
> @amitibo the first case I think is already covered just by defining multiple optimizers, no need of "stem lr", "blocks lr" etc ?

@vfdev-5  Note, that in the first case, the three schedulers ,if called with `save_history=True`, will overwrite the same field, "lr", in `engine.state.param_history`. Therefore there is a need to allow for passing a unique name for each param_scheduler.
 
> The second case is very similar to the first case and just avoids defining multiple optimizers and we can reuse simple engines with a single optimizer.

I think that `ignite` should support the param_groups feature of `pytorch`

**Comment by vfdev-5** [comment]
@amitibo OK, I see, would not be better to check internally whether key exists and add something to avoid overwritting. If user sets the same names (by error, copy-pasting) it will overwrite as well ?

> I think that ignite should support the param_groups feature of pytorch

I agree, this will be a good feature !

**Comment by vfdev-5** [comment]
> Possible solution would be to check whether this is the first time __call__ is used, and if so and the corresponding field already exists, add some random name 'lr (copy ####)'?

@amitibo yes, something like that.
